Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1242,6 +1264,34 @@ void CGContextClearRect(CGContextRef context, CGRect rect) {
     renderTarget->PopAxisAlignedClip();
 }
 
+static bool __CGContextCreateShadowEffect(CGContextRef context, ID2D1DeviceContext* deviceContext, ID2D1Image* inputImage, ID2D1Effect** outShadowEffect) {
+    auto& state = context->CurrentGState();
+    if (std::fpclassify(state.shadowColor.w) != FP_ZERO) {
+        ComPtr<ID2D1Effect> shadowEffect;
+        ComPtr<ID2D1Effect> affineTransformEffect;
+        ComPtr<ID2D1Effect> compositeEffect;
+
+        deviceContext->CreateEffect(CLSID_D2D1Shadow, &shadowEffect);
+        deviceContext->CreateEffect(CLSID_D2D12DAffineTransform, &affineTransformEffect);
+        deviceContext->CreateEffect(CLSID_D2D1Composite, &compositeEffect);
+
+        shadowEffect->SetInput(0, inputImage);
+        shadowEffect->SetValue(D2D1_SHADOW_PROP_COLOR, state.shadowColor);
+        shadowEffect->SetValue(D2D1_SHADOW_PROP_BLUR_STANDARD_DEVIATION, state.shadowBlur);
+
+        affineTransformEffect->SetInputEffect(0, shadowEffect.Get());
+        affineTransformEffect->SetValue(D2D1_2DAFFINETRANSFORM_PROP_TRANSFORM_MATRIX, D2D1::Matrix3x2F::Translation(state.shadowOffset.width, state.shadowOffset.height));
+
+        compositeEffect->SetInputEffect(0, affineTransformEffect.Get());
+        compositeEffect->SetInput(1, inputImage);
```

Comment:
super duper nit: I personally would move the creation and use of composite effect all to the end just so its :

Things with A
Things with B
Things that combine A and B